logins-sql/src/db.rs

```diff
@@ -0,0 +1,789 @@
+    }
+
+    #[inline]
+    pub fn execute_cached(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, true)?)
+    }
+
+    fn do_exec(&self, sql: &str, params: &[(&str, &ToSql)], cache: bool) -> Result<usize> {
+        let res = if cache {
+            self.db.prepare_cached(sql)
+                   .and_then(|mut s| s.execute_named(params))
+        } else {
+            self.db.execute_named(sql, params)
+        };
+        if let Err(e) = &res {
+            warn!("Error running SQL {}. Statement: {:?}", e, sql);
+        }
+        Ok(res?)
+    }
+
+    pub fn query_one<T: FromSql>(&self, sql: &str) -> Result<T> {
+        let res: T = self.db.query_row(sql, &[], |row| row.get(0))?;
+        Ok(res)
+    }
+
+    // The type returned by prepare/prepare_cached are different, but must live
+    // to the end of the function, so (AFAICT) it's difficult/impossible to
+    // remove the duplication between query_row/query_row_cached.
+
+    pub fn query_row<T>(&self, sql: &str, args: &[(&str, &ToSql)], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare(sql)?;
+        let res = stmt.query_named(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+
+    pub fn query_row_cached<T>(&self, sql: &str, args: &[(&str, &ToSql)], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare_cached(sql)?;
+        let res = stmt.query_named(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+}
+
+// login specific stuff.
+
+impl LoginDb {
+
+    fn mark_as_synchronized(&mut self, guids: &[&str], ts: ServerTimestamp) -> Result<()> {
+        util::each_chunk(guids, self.max_var_count, |chunk, _| {
+            self.db.execute(
+                &format!("DELETE FROM {mirror_table} WHERE guid IN ({vars})",
+                         mirror_table = schema::MIRROR_TABLE_NAME,
+                         vars = util::sql_vars(chunk.len())),
+                chunk
+            )?;
+
+            self.db.execute(
+                &format!("
+                    INSERT OR IGNORE INTO {mirror_table} (
+                        {common_cols}, is_overridden, server_modified
+                    )
+                    SELECT {common_cols}, 0, {modified_ms_i64}
+                    FROM {local_table}
+                    WHERE guid IN ({vars})",
+                    common_cols = schema::COMMON_COLS,
+                    mirror_table = schema::MIRROR_TABLE_NAME,
```

Comment:
I have a slight preference for encoding the table names and columns directly in the query, even if they are duplicated; I think it makes grepping and scanning easier. (The BrowserDB code on Android also interpolates constants like this, and it definitely slowed me down understanding the schema. Figuring out where the constant is defined, backtracking from the column getters, and finding all queries mentioning a table was a little cumbersome). But it's a minor nit, and totally your call.

Reply:
Yeah, ios didn't do it, and it's tedious and pointless. I'll take this as a recommendation to move to encoding them directly everywhere.